### ISSUE TITLE
Asynchronous web client sending request to triton server

### ISSUE THREAD

**Issue by yichong96** [triton-inference-server/server#3671]
In the `image_client.py`, the requests that are asynchronous are appended to the `async_requests` list. 

```
 async_requests.append(
       triton_client.async_infer(
       FLAGS.model_name,
       inputs,
       request_id=str(sent_count),
       model_version=FLAGS.model_version,
       outputs=outputs))
```

Once all the requests have been sent, the client later calls the blocking `get_result()` on each `AsyncInferRequest` which blocks the thread.

```
if FLAGS.async_set:
    # Collect results from the ongoing async requests
    # for HTTP Async requests.
    for async_request in async_requests:
        responses.append(async_request.get_result())
```

I am trying to implement a web service using  the `Fast API` framework. Within this service I am making calls to triton server to do inference for each request. I would like to know if there is an `async/await` or `callback` pattern that triton provides so that I can serve multiple requests concurrently while waiting for the inference results of the current request to be done. 

I believe the example provided does not really illustrate the power of async because it is calling the blocking `get_result()` ?

Hope my question was clear !

**Comment by tanmayv25** [comment]
You can use the gRPC endpoint for callback based async operations. Triton python HTTP client is based off on geventhttplib library for http transaction which again uses cooperative scheduling for concurrent operations. Even though light-weight, this has an added restriction for writing code that is greenlet friendly. 

If you still want to use the http endpoint, I would advice using writing your own implementation using asyncio/aiohttp modules may be. If you look at the [client code](https://github.com/triton-inference-server/client/blob/main/src/python/library/tritonclient/http/__init__.py) it is just a thin wrapper around triton API.

**Comment by yichong96** [comment]
I see.. If i am not wrong the greenlets will execute HTTP requests concurently but the main thread will still block until all the requests have been completed am I right ?

**Comment by tanmayv25** [comment]
It would be clearer to think the HTTP requests will be sent cooperatively. When the first async request is sent, the [geventhttpclient](https://github.com/geventhttpclient/geventhttpclient) will initiate the HTTP request and when being blocked by network I/O, it will pass the control to send another async request, same is true for other requests until you call the `get_result()`. In which case, it will start waiting for the response to arrive for that specific request and block on it. This level of cooperation is required to proceed with the event loop. 

So, yes overall it appears to be sending requests asynchronously and then blocking on getting result. You don't have to wait for receiving all responses before proceeding. You can consume the result from first response and do whatever post-processing you want, then call get_result on next request only when required..

**Comment by yichong96** [comment]
@tanmayv25 thank you for the explanation ! 

I am thinking if the async/await pattern is superior because I can `await` for all asynchronous triton requests that I send to be ready before postprocessing the results all at once. While awaiting for all triton requests to be completed, the main thread can serve other HTTP clients. 

Do you think this is better ?

**Comment by tanmayv25** [comment]
Yes.. depending upon your use-case if your main thread is doing some additional work instead of blocking, then it should be good. The advantage of greenlets based design is that you have more control over what you want to do at the given time. You can read more about greenlets here: https://greenlet.readthedocs.io/en/latest/.
You can have one thread that keeps on issuing request to the server using async_infer and store the async_request in a sync queue. The other thread can keep consuming the async_request and retrieve the results. There are several possible design. The best one depends upon your use case.

**Comment by fifofonix** [comment]
FWIW in our implementation of TensorFlow ModelServer (very similar to Triton except I'm not sure it has the same backends concept) we use [gRPC AsyncIO](https://grpc.github.io/grpc/python/grpc_asyncio.html) when ripping through video frames and farming out to a multiple model server instances.  The way we use it I think we get all the advantages of the greenlet approach but with a possibly more intuitive/standard asyncio model?

I now find myself looking to integrate TritonServer in parallel to the above and would ideally want to implement within the same asyncio model.  It is not yet clear to me whether this is possible (and I won't be looking at this until the New Year at this point) but certainly if it is possible an asyncio based example would make a great addition to the examples package.  I'm not familiar with greenlets but high level documentation at least suggests this also does not use threads - which would be good - because threads would not mix well with the asyncio implementation we have already as I understand it.

**Comment by alexzubiaga** [comment]
Here is an example how to make async requests using the triton client to generate the request contents and httpx to handle the async communication. Please note that it's a little bit hacky as it is accessing an internal method of the triton client and may break with future versions.

```
from typing import List, Optional
from urllib.parse import quote

import httpx
import tritonclient
from tritonclient.http import InferResult, InferInput, InferRequestedOutput


async def infer(
        base_url: str, model_name: str, model_version: str,
        inputs: List[InferInput], outputs: List[InferRequestedOutput],
        timeout: Optional[int] = None) -> InferResult:
 
    request_body, json_size = tritonclient.http._get_inference_request(
        inputs=inputs, request_id='', outputs=outputs,
        sequence_id=0, sequence_start=0, sequence_end=0,
        priority=0, timeout=timeout)

    headers = {}
    if json_size is not None:
        headers["Inference-Header-Content-Length"] = str(json_size)
    request_uri = "v2/models/{}/versions/{}/infer".format(quote(model_name), model_version)

    async with httpx.AsyncClient(base_url=base_url, headers=headers, timeout=120) as client:
        triton_response = await client.post(url=request_uri, data=request_body)
    tritonclient.http._raise_if_error(triton_response)

    inference = InferResult.from_response_body(
        response_body=triton_response.content,
        verbose=False,
        header_length=triton_response.headers['Inference-Header-Content-Length'])
    return inference
```

**Comment by fifofonix** [comment]
Thanks @alexzubiaga.  As it happens, as of a few hours ago, I got the async approach working (via grpc) but without needing to use an internal method.  Basically, I imported the aio submodule in grpc and used that to establish the channel.  

This seems to work just fine allowing me to use a producer/consumer paradigm where a number of consumers results in concurrent grpc requests.  These have proven to be handled just fine for the past several hours although I haven't yet scaled to a farm of Triton servers yet.

Skeleton code:

```python

import asyncio

import grpc
from grpc import aio

from tritonclient.grpc import service_pb2 as predict_pb2
from tritonclient.grpc import service_pb2_grpc as prediction_service_pb2_grpc

# This called from an asyncio event loop as normal.
def async my_infer(image_np):
    async with aio.insecure_channel("localhost:8001") as channel:
        stub = prediction_service_pb2_grpc.GRPCInferenceServiceStub(channel)
        # Setup the request with input/outputs as normal (not shown to simplify).
        # ...
        response = await stub.ModelInfer(request)
    return response
```

**Comment by narolski** [comment]
@fifofonix based on your proposal I've implemented an async grpc infer method:
```
    async def get_inference_results_grpc(self, input_image: np.ndarray) -> np.array:
        """
        Returns the CRAFT inference result for an `input_image`
        by performing the request to the Nvidia Triton
        server via gRPC.

        Awaitable code implemented based on:
        https://github.com/triton-inference-server/server/issues/3671.

        Parameters
        ----------
        input_image : np.ndarray
            Image to be processed by the CRAFT scene text
            detection model.

        Returns
        -------
        np.array
            Array of inference results.

        """
        input_image = detection.compute_input(input_image)
        input = grpcclient.InferInput(
            name=TRITON_MODEL_INPUT_NAME,
            shape=[1, input_image.shape[0], input_image.shape[1], input_image.shape[2]],
            datatype="FP32",
        )
        output = grpcclient.InferRequestedOutput(TRITON_MODEL_OUTPUT_NAME)
        try:
            input.set_data_from_numpy(np.array([input_image]))
        except InferenceServerException as exc:
            logger.error("Cannot set input data for the tensor", e=str(exc))
            raise HTTPException(status_code=400, detail="Invalid CRAFT Input Data")

        try:
            async with aio.insecure_channel(TRITON_GRPC_SERVICE) as channel:
                stub = prediction_service_pb2_grpc.GRPCInferenceServiceStub(channel)
                request = _get_inference_request(
                    model_name=TRITON_MODEL_NAME,
                    model_version=TRITON_MODEL_VERSION,
                    inputs=[input],
                    outputs=[output],
                    timeout=TRITON_GRPC_TIMEOUT,
                    request_id="",
                    sequence_id=0,
                    sequence_start=0,
                    sequence_end=0,
                    priority=0,
                )
                response = await stub.ModelInfer(request)
        except grpc.aio.BaseError as exc:
            logger.error("Error establishing gRPC connection to Triton", e=str(exc))
            raise HTTPException(status_code=500, detail="gRPC Connection Error")
        except InferenceServerException as exc:
            logger.error("Error handling inference request via gRPC to Triton", e=str(exc))
            raise HTTPException(status_code=500, detail="gRPC Inference Error")

        result = InferResult(response)
        result = result.as_numpy(name=TRITON_MODEL_OUTPUT_NAME)

        return result
```

However, it introduces a memory leak - it seems that the memory is not freed when (when we fetch the response using `await stub.ModelInfer(request)`, or when we create `input`, `output` objects.

Have you faced this issue by any chance? Did you manage to find a possible fix? I'd be much obliged 🙏

**Comment by fifofonix** [comment]
We are not experiencing a memory issue with this strategy.  Our processes are running as containers and we see them hitting a plateau of memory utilization which is more or less steady across many 100Ks inferences.  The containers are using a provider/consumer asyncio paradigm and each have 8 or so concurrent inferences on the go at any one time.  I have scanned your code visually and unfortunately I don't see anything obviously wrong.

We're using python 3.6.9 with:
grpcio==1.34.1
tritonclient==2.17.0

**Comment by piekey1994** [comment]
> @fifofonix thank you, sadly I've had to revert to the synchronous gRPC inference. I am running the code on Python 3.10.2, so this might be one of the reasons.
@fifofonix We use grpc synchronized + loop.run_in_executor or await stub.ModelInfer(request) mentioned above, there will be memory leaks. However, this problem does not occur when using loop.run_in_executor + httpclient.

**Comment by fifofonix** [comment]
@MatthieuTPHR you can certainly overwhelm the model server in which case it will indicate it is unavailable for new requests so this is expected.  You can dynamically alter the depth of your consumer queue in response to server errors like this so that you feed the model server at an optimal rate and maximize throughput.

We do use this strategy in production and it is very stable.  For 98% of the time we are operating with zero GRPC errors with all responses collected prior to our timeout.  However, the environment is dynamic and during periods of scale up / scale down when load balancers are catching up with model servers going on/offline you have to expect some errors and be prepared to resubmit.

We are actually using both Triton and TensorFlow model server in our setup although we plan on running an experiment to move solely to Triton for its multi-model-type flexibility.  

Adding to prior comments from @piekey1994 and @narolski regarding memory usage, we can confirm that the asyncio model does balloon with deferred garbage collection.  To address this we have simply provided a maximum memory limit on our containers - we see a balloon to this point and then python performs garbage collection.  I've read some threads on this and there are other strategies for releasing memory sooner but we have not found a need to pursue them.

**Comment by manastahir** [comment]
I don't know if this is still relevant, but I found a good and simple example of making async requests using GRPC client and `asyncio` libraray https://github.com/triton-inference-server/client/blob/main/src/python/examples/simple_grpc_aio_infer_client.py

**Comment by neilbhutada** [comment]
There is a method called: async_infer. 

This is an example on how to use it:

```
import asyncio
import tritonserver
server = tritonserver.Server(model_repository="/workspace/models")
model = server.model("<model repo>")

async def send_requests(input_strings):
    async for responses in  model.async_infer(inputs={"string_input":input_strings}): 
        return [json.loads(answer) for answer in responses.outputs["string_output"].to_string_array().tolist()]

async def send_all(x):
    tasks = [send_requests(input) for input in x]
     results = await asyncio.gather(*tasks)
     return results

asyncio.run(send_all(<your inputs>))

```
